File: pset2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Prototypes
int count_sentences(string t);
int count_words(string t);
int count_letters(string t);

int main(void)
{
    // Prompt for text
    string text = get_string("Text: ");

    // Count sentences, words, and letters
    int sentences = count_sentences(text);
    int words = count_words(text);
    int letters = count_letters(text);

    // Compute the Coleman-Liau index.
    // index = 0.0588 * L - 0.296 * S - 15.8

    float l_avg = (letters / (float) words) * 100;
    float s_avg = (sentences / (float) words) * 100;

    int index = round((0.0588 * l_avg) - (0.296 * s_avg) - 15.8);

    // Print grade
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

// Count sentences
int count_sentences(string t)
{
    int s = 0; // Sentences

    for (int i = 0, l = strlen(t); i < l; i++)
    {
        // Sentences end with {. || ! || ?}
        if (t[i] == '.' || t[i] == '!' || t[i] == '?')
        {
            s++;
        }
    }
    return s;
}

// Count words
int count_words(string t)
{
    int w = 0; // Words

    for (int i = 0, l = strlen(t); i < l; i++)
    {
        // Add new word if found space
        if (isspace(t[i]))
        {
            w++;
        }
    }
    return ++w;
}

// Count letters
int count_letters(string t)
{
    int lt = 0; // Letters

    for (int i = 0, l = strlen(t); i < l; i++)
    {
        // Alphabetical letters
        if (isalpha(t[i]))
        {
            lt++;
        }
    }
    return lt;
}

File: pset6/sentimental-readability/readability.py
def main():
    # Prompt user for text
    text = input("Text: ")

    # Count sentences, words, letters
    counts = get_counts(text)

    sentences = counts["sentences"]
    words = counts["words"]
    letters = counts["letters"]

    # Calculate Coleman-Liau index formula
    # index = 0.0588 * L - 0.296 * S - 15.8

    l_avg = (letters / words) * 100
    s_avg = (sentences / words) * 100

    index = int(round((0.0588 * l_avg) - (0.296 * s_avg) - 15.8))

    print_grade(index)


# Get text counts
def get_counts(text):
    counts = {
        "sentences": 0,
        "words": 0,
        "letters": 0
    }

    for c in text:  # sentences
        if c in ["!", "?", "."]:
            counts["sentences"] += 1
        elif c == " ":  # words
            counts["words"] += 1
        elif c.isalpha():  # letters
            counts["letters"] += 1
    
    # Last word in a text
    counts["words"] += 1

    return counts


# Print reading grade according to index
def print_grade(index):
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


if __name__ == "__main__":
    main()